File: backend/app/services/agent_loop.py
from __future__ import annotations

import json
from dataclasses import dataclass, field

import anthropic

from app.config import settings
from app.services.context_loader import TenantContext
from app.services.system_prompt import build_system_prompt
from app.services.tool_handlers import AGENT_TOOLS, execute_tool

MAX_ITERATIONS = 8


@dataclass
class AgentResult:
    final_message: str
    tools_used: list[str] = field(default_factory=list)
    high_severity_actions: list[str] = field(default_factory=list)
    intent_classification: str = "general"
    confidence_score: float = 0.8


async def run_agent_loop(user_message: str, ctx: TenantContext) -> AgentResult:
    client = anthropic.Anthropic(api_key=settings.ANTHROPIC_API_KEY)

    system_prompt = build_system_prompt(ctx)

    # Seed conversation history from DB (last 10 messages, chronological)
    messages: list[dict] = [
        {
            "role": "user" if conv.direction == "inbound" else "assistant",
            "content": conv.message_body,
        }
        for conv in ctx.recent_conversations
    ]
    messages.append({"role": "user", "content": user_message})

    tools_used: list[str] = []
    high_severity_actions: list[str] = []
    final_message = ""
    iteration = 0

    while iteration < MAX_ITERATIONS:
        iteration += 1

        response = client.messages.create(
            model="claude-sonnet-4-5",
            max_tokens=1024,
            system=system_prompt,
            tools=AGENT_TOOLS,  # type: ignore[arg-type]
            messages=messages,
        )

        # Append assistant turn
        messages.append({"role": "assistant", "content": response.content})

        if response.stop_reason == "end_turn":
            for block in response.content:
                if block.type == "text":
                    final_message = block.text
                    break
            break

        if response.stop_reason == "tool_use":
            tool_results = []

            for block in response.content:
                if block.type != "tool_use":
                    continue

                tool_name = block.name
                tool_input = block.input or {}
                tools_used.append(tool_name)

                try:
                    result = await execute_tool(tool_name, tool_input, ctx)
                except Exception as exc:
                    result_content = json.dumps({"error": str(exc)})
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": block.id,
                        "content": result_content,
                        "is_error": True,
                    })
                    continue

                if result.is_high_severity and result.landlord_notification_message:
                    high_severity_actions.append(result.landlord_notification_message)

                result_content = json.dumps(
                    result.data if result.success else {"error": result.error}
                )
                tool_results.append({
                    "type": "tool_result",
                    "tool_use_id": block.id,
                    "content": result_content,
                    "is_error": not result.success,
                })

            messages.append({"role": "user", "content": tool_results})
            continue

        # Unexpected stop reason — use any text available
        for block in response.content:
            if block.type == "text":
                final_message = block.text
                break
        break

    if not final_message:
        final_message = "I have processed your request. Please let me know if you need anything else."

    intent = _classify_intent(user_message, tools_used)
    confidence = 0.95 if tools_used else 0.8

    return AgentResult(
        final_message=final_message,
        tools_used=tools_used,
        high_severity_actions=high_severity_actions,
        intent_classification=intent,
        confidence_score=confidence,
    )


def _classify_intent(message: str, tools_used: list[str]) -> str:
    if "issue_legal_notice" in tools_used:
        return "legal_response"
    if "schedule_maintenance" in tools_used:
        return "maintenance"
    if "get_rent_status" in tools_used:
        return "finance"
    if "update_escalation_level" in tools_used:
        return "escalation"

    lower = message.lower()
    if any(w in lower for w in ("rent", "pay", "arrear")):
        return "finance"
    if any(w in lower for w in ("fix", "broken", "repair", "leak", "heat")):
        return "maintenance"
    if any(w in lower for w in ("lease", "contract", "renew")):
        return "lease_query"
    if any(w in lower for w in ("notice", "evict", "legal")):
        return "legal_response"
    return "general"

File: frontend/lib/agent/agentLoop.ts
import Anthropic from '@anthropic-ai/sdk'
import { TenantContext } from './contextLoader'
import { buildSystemPrompt } from './systemPrompt'
import { AGENT_TOOLS, ToolName, ToolInput } from './tools'
import { executeTool, ToolResult } from './toolHandlers'

const anthropic = new Anthropic({
  apiKey: process.env.ANTHROPIC_API_KEY!,
})

const MAX_ITERATIONS = 8 // Safety limit for tool-calling loops

export interface AgentResult {
  finalMessage: string
  toolsUsed: string[]
  highSeverityActions: string[]
  intentClassification: string
  confidenceScore: number
}

export async function runAgentLoop(
  userMessage: string,
  ctx: TenantContext
): Promise<AgentResult> {
  const systemPrompt = buildSystemPrompt(ctx)

  // Seed the messages array with the full conversation history + new message
  const messages: Anthropic.MessageParam[] = [
    ...buildConversationHistory(ctx),
    { role: 'user', content: userMessage },
  ]

  const toolsUsed: string[] = []
  const highSeverityActions: string[] = []
  let iteration = 0
  let finalMessage = ''

  while (iteration < MAX_ITERATIONS) {
    iteration++

    const response = await anthropic.messages.create({
      model: 'claude-sonnet-4-5',
      max_tokens: 1024,
      system: systemPrompt,
      tools: AGENT_TOOLS,
      messages,
    })

    // Append assistant response to messages
    messages.push({ role: 'assistant', content: response.content })

    if (response.stop_reason === 'end_turn') {
      // Extract the final text response
      const textBlock = response.content.find((b) => b.type === 'text')
      finalMessage = textBlock ? (textBlock as Anthropic.TextBlock).text : ''
      break
    }

    if (response.stop_reason === 'tool_use') {
      // Execute all tool calls in this response
      const toolUseBlocks = response.content.filter(
        (b): b is Anthropic.ToolUseBlock => b.type === 'tool_use'
      )

      const toolResults: Anthropic.ToolResultBlockParam[] = []

      for (const toolUse of toolUseBlocks) {
        const toolName = toolUse.name as ToolName
        toolsUsed.push(toolName)

        let result: ToolResult
        try {
          result = await executeTool(toolName, toolUse.input as ToolInput[typeof toolName], ctx)
        } catch (err) {
          result = {
            success: false,
            error: `Tool execution failed: ${err instanceof Error ? err.message : String(err)}`,
          }
        }

        if (result.isHighSeverity && result.landlordNotificationMessage) {
          highSeverityActions.push(result.landlordNotificationMessage)
        }

        toolResults.push({
          type: 'tool_result',
          tool_use_id: toolUse.id,
          content: JSON.stringify(result.success ? result.data : { error: result.error }),
          is_error: !result.success,
        })
      }

      // Feed tool results back to Claude
      messages.push({ role: 'user', content: toolResults })
      continue
    }

    // Unexpected stop reason — break and use whatever text we have
    const textBlock = response.content.find((b) => b.type === 'text')
    finalMessage = textBlock ? (textBlock as Anthropic.TextBlock).text : 'I was unable to process your request. Please try again.'
    break
  }

  if (!finalMessage) {
    finalMessage = 'I have processed your request. Please let me know if you need anything else.'
  }

  // Classify the intent from the conversation for logging
  const intentClassification = classifyIntent(userMessage, toolsUsed)

  return {
    finalMessage,
    toolsUsed,
    highSeverityActions,
    intentClassification,
    confidenceScore: toolsUsed.length > 0 ? 0.95 : 0.8,
  }
}

function buildConversationHistory(ctx: TenantContext): Anthropic.MessageParam[] {
  // Convert stored conversations to Claude message format (last 10, already in chronological order)
  return ctx.recentConversations.map((conv) => ({
    role: conv.direction === 'inbound' ? 'user' : 'assistant',
    content: conv.message_body,
  }))
}

function classifyIntent(message: string, toolsUsed: string[]): string {
  if (toolsUsed.includes('issue_legal_notice')) return 'legal_response'
  if (toolsUsed.includes('schedule_maintenance')) return 'maintenance'
  if (toolsUsed.includes('get_rent_status')) return 'finance'
  if (toolsUsed.includes('update_escalation_level')) return 'escalation'

  const lower = message.toLowerCase()
  if (lower.includes('rent') || lower.includes('pay') || lower.includes('arrear')) return 'finance'
  if (lower.includes('fix') || lower.includes('broken') || lower.includes('repair') || lower.includes('leak') || lower.includes('heat')) return 'maintenance'
  if (lower.includes('lease') || lower.includes('contract') || lower.includes('renew')) return 'lease_query'
  if (lower.includes('notice') || lower.includes('evict') || lower.includes('legal')) return 'legal_response'
  return 'general'
}
